Epics in docker

gustavosr8 opened the pull request.

gustavosr8 committed:
Use standardized containerization scheme.

ericonr commented on docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+      args:
+        REPONAME: ipmi-mgr-epics-ioc
+        RUNDIR: /opt/ipmi-mgr-epics-ioc/iocBoot/iocIpmiMgr
+          #ENTRYPOINT: ./YOUR_ENTRYPOINT
+    command: ./runIpmiMgr.sh -P HOMOLOG -R DE22 -i de-22rabpm-co-cratectrl
```
Leftover command.

And usually we don't leave command with something to be used for debugging, but I don't know if it's a bad idea...

I think we should cleanup the scripts used here, though. The containers for the machine usually require a complicated shell script mess, when they could just use environment variables in the docker compose file, which have the advantage of being self-documenting (we can discuss if that pattern makes more sense in an IOC meeting); then, it also allows us to directly use `lnls-run`.

henriquesimoes replied:
Do you think we may follow something similar to what we did for RFFE IOC?

That is, https://github.com/lnls-dig/rffe-epics-ioc/commit/655c72f921ed841c9aaae89ee8d1dfd908140073 (simplified by https://github.com/lnls-dig/rffe-epics-ioc/commit/286286f6c084c1ac31949e48f714d631886fafe2).

ericonr replied:
I don't think so, unless all you want is an example config... The actual deployment will be in the internal docker-stacks repository, since this service will run on the cluster and not on the crates' CPUs (we want to be able to remotely restart the CPU)

ericonr commented:
This PR is missing the changes to the `.gitmodules` file

henriquesimoes commented on install/Makefile:
```diff
@@ -1,33 +0,0 @@
-uninstall: scripts_uninstall
-	$(TOP)/install/uninstall.sh ${INSTALL_PREFIX}/${IOC_NAME} $(TOP)
-
-scripts_install:
-	$(MAKE) -C ${SCRIPTS_FOLDER} install
-
-scripts_uninstall:
-	$(MAKE) -C ${SCRIPTS_FOLDER} uninstall
```
You remove these rules, but do not remove their target folder (i.e. `scripts`).

IMHO, this removal could be done in the same commit that adds the deploy compose file if we decide to follow that approach. This should make clearer that `compose.yml` substitute what systemd services were previously doing (with the except of actually launching the containers).

henriquesimoes reviewed (changes requested):
Maybe we can add the CI as well. Not sure if in PR, though.

henriquesimoes commented:
I tried to use a non-intuitive behavior from GitHub review interface, but it didn't behave as I expected, and it dropped some of my comments. Here they are:

## Commit aadd41baebffb947819a46426e85ef4c296c8ba8

Suggestion: instead of using "Add the epics-in-docker submodule and the docker-compose.yml file." as title, which describes what `git show --stat aadd41bae` already tells us, we can use a shorter and still descriptive title "Use standardized containerization scheme".

## Commit 79b0bb2994574c942faf7beb33fa26a38bb97b1b

Message suggestion: move the justification clause to the message body. IMO, the point here is that out-of-source installation should be performed by the epics-in-docker containerization layer, even though it is currently not the case.

## Commit c9e85f4289ab124aa3152d3066bd2808478aba5e

> For this reason, we need to install all the modules that we need at runtime.

That's not quite the case. We need to install locally all support module _artifacts_ (not the modules themselves) that are used by the IOC at runtime.

gustavosr8 committed:
Remove initialization scripts.

The `lnls-run` entrypoint launch the `st.cmd` command under procServ.

ericonr commented:
Does this depend on the changes we talked about for the ipmiComm module?

henriquesimoes commented on docker-compose.yml:
```diff
@@ -0,0 +1,17 @@
+      P: ${P}
+      R: ${R}
+      IPADDR: ${IPADDR}
+
```
Nit: useless extra empty line.

henriquesimoes reviewed (approved):
Seems good to me.

> Does this depend on the changes we talked about for the ipmiComm module?

I don't know what you are talking about here, so not taking this into account in this last review. ;)

henriquesimoes reviewed (changes requested):
I noticed now that there is something that actually blocks this: a release commit after v0.8.1 should be used in the `docker` submodule.

As of now, we do not have a release that can be used, though.

gustavosr8 committed:
Install asynRecord locally.

After build, epics_in_docker deletes the directories with the support
modules. For this reason, we need to install locally all support module
artifacts that are used by the IOC at runtime.

henriquesimoes commented on ipmiMgrApp/Db/afcv3.substitutions:
```diff
@@ -41,22 +41,22 @@ file "power_mngr.db"
-            { TempClkSwitch   , 3           , 1       , $(fruid)     , $(id)  }
-            { TempDCDC        , 4           , 1       , $(fruid)     , $(id)  }
-            { TempRAM         , 5           , 1       , $(fruid)     , $(id)  }
+    pattern { attr            , sensinst    , type    , fruid        , prefix   , aliasprefix }
+            { FMC1Volt12V     , 1           , 2       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC1VoltVADJ    , 2           , 2       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC1Volt3V3     , 3           , 2       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC2Volt12V     , 4           , 2       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC2VoltVADJ    , 5           , 2       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC2Volt3V3     , 6           , 2       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC1Curr12V     , 1           , 3       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC1CurrVADJ    , 2           , 3       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC1Curr3V3     , 3           , 3       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC2Curr12V     , 4           , 3       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC2CurrVADJ    , 5           , 3       , $(fruid)     , $(id)    , alias_$(id) }
+            { FMC2Curr3V3     , 6           , 3       , $(fruid)     , $(id)    , alias_$(id) }
+            { TempFPGA        , 1           , 1       , $(fruid)     , $(id)    , alias_$(id) }
+            { TempUC          , 2           , 1       , $(fruid)     , $(id)    , alias_$(id) }
+            { TempClkSwitch   , 3           , 1       , $(fruid)     , $(id)    , alias_$(id) }
+            { TempDCDC        , 4           , 1       , $(fruid)     , $(id)    , alias_$(id) }
+            { TempRAM         , 5           , 1       , $(fruid)     , $(id)    , alias_$(id) }
```
Please, document in the commit message why `noalias=#` is not an option if we use the `include` strategy.

henriquesimoes commented on ipmiMgrApp/Db/Makefile:
```diff
@@ -10,7 +10,7 @@ include $(TOP)/configure/CONFIG
 # Create and install (or just install) into <top>/db
 # databases, templates, substitutions like this
-
+USR_DBFLAGS += -I$(IPMICOMM)/Db
```
This should be a standalone commit, referencing the commit it fixes. Also mention that it can be removed as soon as a new release comes up.

henriquesimoes replied:
I don't like very much having a commit hash in the header line. A commonly used pattern is to add a trailer specifying the commit that is fixed. For instance, the message could be something like:

```
Search for database files in ipmiComm Db

We use some databases from the ipmiComm module that are not available in
the `db` directory, only in the `Db`. This will no longer be true as
soon as a new version containing the export [1] is released in
epics-in-docker.

Fixes: 14a436eaad83 (Change the locally defined Db's to use aliases from the ipmiComm db's)

[1]: https://github.com/slac-epics-modules/ipmiComm/commit/5e1939de5b14e49cb0c2e79f3df6ee8b0a94704c
```

Adding the message right after the hash is [a practice highly encouraged by Linux kernel community][1] and also commonly seen in Git project itself.

Also note that in my message suggestion, I've added enough information for someone to evaluate if the change should be reverted. IMO, that's better than having it loosely defined.

[1]: https://www.kernel.org/doc/html/v6.10/process/submitting-patches.html#describe-your-changes

henriquesimoes commented on iocBoot/iocIpmiMgr/stIpmiMgr.cmd:
```diff
@@ -1,6 +1,5 @@
 < envPaths
 epicsEnvSet("TOP", "../..")
-< IpmiMgr.config
```
This should be in the same commit the file itself is removed, that is 639d3e7f7311751d10b61cc0a7dc9eaef0374eb6, not this one.

henriquesimoes replied:
This is still in "Remove install scripts" not in "Remove initialization scripts". The same for `PORT` definition.

henriquesimoes commented on iocBoot/iocIpmiMgr/stIpmiMgr.cmd:
```diff
@@ -11,6 +10,7 @@ ipmiMgr_registerRecordDeviceDriver pdbbase
 var dbRecordsOnceOnly 1
 
 ## Initialize connection to MCH
+epicsEnvSet("PORT", "MCHLink")
```
Same here

henriquesimoes commented on docker:
The checked out commit is wrong. It should be the one from the release.

henriquesimoes commented on ipmiMgrApp/Db/Makefile:
```diff
@@ -10,7 +10,6 @@ include $(TOP)/configure/CONFIG
 #----------------------------------------------------
 # Create and install (or just install) into <top>/db
 # databases, templates, substitutions like this
-
```
Nit: unrelated change; I'd drop this.

gustavosr8 committed:
Remove install scripts.

Out-of-source installation should be performed by the epics-in-docker
containerization layer.

henriquesimoes reviewed (approved):
It seems good for me. Let's see what @ericonr thinks about it.

ericonr commented:
Couldn't we import the ipmiComm fix into epics-in-docker to avoid having to carry the last commit here?

gustavosr8 commented:
> Couldn't we import the ipmiComm fix into epics-in-docker to avoid having to carry the last commit here?

The ipmiComm fix is in an [pre-release branch](https://github.com/slac-epics-modules/ipmiComm/tree/pre-release), so IMO we can wait till it is officially release to import into epics-in-docker. Otherwise it will be also an temporary commit there.

ericonr commented:
Depends on: https://github.com/cnpem/epics-in-docker/pull/73

henriquesimoes commented on docker:
It should now use a yet to be release version, containing new changes.

gustavosr8 committed:
Add substitutions for `aliasprefix` macro.

Set `noalias=#` is not an option because the variable is automatically
set as `noalias=''` when we include the db.
